File: Solutions/java/y2015/Solution02.java
package Solutions.java.y2015;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import Solutions.java.Solution;

public class Solution02 implements Solution {

    @Override
    public void Part1(String inputFile) {
        Path filename = Path.of(inputFile);

        try {
            var total = Files.readAllLines(filename)
            .stream()
            .map(line -> { return new Dimensions(line); })
            .mapToInt(gift -> { return gift.Surface() + gift.MinSurface(); })
            .sum();

            System.out.printf("The elves need to order %d square feet of wrapping paper\n", total);
        }
        catch(IOException exception) {

        }
    }

    @Override
    public void Part2(String inputFile) {
        Path filename = Path.of(inputFile);

        List<Dimensions> gifts = new ArrayList<Dimensions>();

        try {
            gifts = Files.readAllLines(filename)
                .stream()
                .map(line -> { return new Dimensions(line); })
                .collect(Collectors.toList());
        }
        catch(IOException exception) {

        }
        
        var total = 0;
        for (Dimensions gift : gifts) {
            int[] arr = { gift.l, gift.w, gift.h };

            var sum = Arrays.stream(arr)
                .sorted()
                .map(x -> { return x * 2; })
                .limit(2)
                .sum();
               
            total += gift.Cubed() + sum;
        }
        System.out.printf("The elves need to order %d feet of ribbon.\n", total);
    }

    public class Dimensions
    {
        public int l;
        public int w;
        public int h;

        public Dimensions(String dimensions)
        {
            var splitted = dimensions.split("x");
            this.l = Integer.parseInt(splitted[0]);
            this.w = Integer.parseInt(splitted[1]);
            this.h = Integer.parseInt(splitted[2]);
        }

        public int MinSurface() {
            return Math.min(Math.min(l * w, w * h), h * l);
        }
        public int Surface() {
            return 2 * l * w + 2 * w * h + 2 * h * l;
        }
        public int Cubed() {
            return l * w * h;
        }
        // public (int, int) ShortestPerimeter()
        // {
        //     //new[] {l, w, h}.Order().Take(2);
        //     var v1 = l * w;
        //     var v2 = w * h;
        //     var v3 = h * l;

        //     var faces = new[] { v1, v2, v3 };
        //     return (faces.Order().Take(1).Single(), faces.Order().Skip(1).Take(1).Single());
        // }
    }    
}

File: Solutions/csharp/y2015/Solution02.cs
namespace AdventOfCode.Y2015;

[Solution(2015, 2)]
public class Solution02
{
    [Part1]
    public void Part1(string filename)
    {
        var data = File.ReadAllLines(filename).Select(line => new Dimensions(line));
        int total = data.Sum(gift => gift.Surface() + gift.MinSurface());
        
        Console.WriteLine($"The elves need to order \"{total}\" square feet of wrapping paper");
    }

    [Part2]
    public void Part2(string filename)
    {
        var data = File.ReadAllLines(filename).Select(line => new Dimensions(line));
        var total = data.Sum(gift =>
        {
            return new[] { gift.l, gift.w, gift.h }.Order().Take(2).Sum(x => x * 2) + gift.Cubed();
        });
        
        Console.WriteLine($"The elves need to order \"{total}\" feet of ribbon.");
    }

    private record struct Dimensions
    {
        public Dimensions(string dimensions)
        {
            var splitted = dimensions.Split("x");
            l = int.Parse(splitted[0]);
            w = int.Parse(splitted[1]);
            h = int.Parse(splitted[2]);
        } 

        public int l, w, h;
        public int MinSurface() => Math.Min(Math.Min(l * w, w * h), h * l);
        public int Surface() => 2 * l * w + 2 * w * h + 2 * h * l;
        public int Cubed() => l * w * h;
        public (int, int) ShortestPerimeter()
        {
            //new[] {l, w, h}.Order().Take(2);
            var v1 = l * w;
            var v2 = w * h;
            var v3 = h * l;

            var faces = new[] { v1, v2, v3 };
            return (faces.Order().Take(1).Single(), faces.Order().Skip(1).Take(1).Single());
        }
    }
}
